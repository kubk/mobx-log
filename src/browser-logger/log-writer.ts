export type LogWriter = {
  write: (...messages: unknown[]) => void;
};

export class DefaultLogWriter implements LogWriter {
  write(...messages: unknown[]) {
    console.log(...messages);
  }
}

export class CollectingLogWriter implements LogWriter {
  history: string[] = [];

  write(...messages: unknown[]) {
    const strings = messages.map((message): string =>
      (message as any).toString()
    );
    this.history.push(strings.join(' '));
  }

  clear() {
    this.history = [];
  }
}
