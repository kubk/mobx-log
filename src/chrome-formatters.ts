import {
  isObservableArray,
  isObservableMap,
  isObservableObject,
  isObservableSet,
  ObservableMap,
  ObservableSet,
  toJS,
} from 'mobx';

type Css = { style: string };
type Chunk = [string, Css, string];
type Header = [string, Chunk, [string, string]];
type Body = [string, Css, ...any[]];

export type ChromeFormatter<T> = {
  header: (argument: T) => Header | null;
  hasBody: (argument: T) => boolean;
  body: (argument: T) => Body;
};

const styles = {
  mobxName: { style: 'color: rgb(232,98,0)' },
  nullName: { style: 'color: #777' },
  list: {
    style:
      'list-style-type: none; padding: 0; margin: 0 0 0 12px; font-style: normal',
  },
};

const reference = (object: unknown) => {
  if (typeof object === 'undefined' || typeof object === null) {
    return ['span', styles.nullName, typeof object];
  }
  return ['object', { object }];
};

const renderIterableHeader = (
  iterable:
    | ArrayLike<unknown>
    | ObservableMap<unknown, unknown>
    | ObservableSet<unknown>,
  name: string
): Header => {
  const count =
    isObservableMap(iterable) || isObservableSet(iterable)
      ? iterable.size
      : iterable.length;

  return ['span', ['span', styles.mobxName, name], ['span', `[${count}]`]];
};

const renderIterableBody = (
  collection:
    | ArrayLike<unknown>
    | ObservableMap<unknown, unknown>
    | ObservableSet<unknown>
    | {},
  mapper: (val: [string, unknown]) => unknown
): Body => {
  if (isObservableMap(collection) || isObservableSet(collection)) {
    // @ts-ignore
    const children = Array.from(toJS(collection)).map(mapper);
    return ['ol', styles.list, ...children];
  }

  const children = Object.entries(toJS(collection)).map(mapper);
  return ['ol', styles.list, ...children];
};

export class ObjectFormatter implements ChromeFormatter<{}> {
  header(argument: {}): Header | null {
    return isObservableObject(argument)
      ? renderIterableHeader(Object.keys(argument), 'Object')
      : null;
  }

  hasBody(argument: {}): boolean {
    return Object.keys(argument).length > 0;
  }

  body(argument: {}) {
    return renderIterableBody(argument, ([key, value]) => [
      'li',
      {},
      reference(key),
      ': ',
      reference(value),
    ]);
  }
}

export class ArrayFormatter implements ChromeFormatter<ArrayLike<unknown>> {
  header(argument: ArrayLike<unknown>): Header | null {
    return isObservableArray(argument)
      ? renderIterableHeader(argument, 'Array')
      : null;
  }

  hasBody(argument: ArrayLike<unknown>): boolean {
    return argument.length > 0;
  }

  body(argument: ArrayLike<unknown>): Body {
    return renderIterableBody(argument, ([_, value]) => [
      'li',
      reference(value),
    ]);
  }
}

export class MapFormatter implements ChromeFormatter<Map<unknown, unknown>> {
  header(argument: Map<unknown, unknown>): Header | null {
    return isObservableMap(argument)
      ? renderIterableHeader(argument, 'Map')
      : null;
  }

  hasBody(argument: Map<unknown, unknown>): boolean {
    return argument.size > 0;
  }

  body(argument: Map<unknown, unknown>): Body {
    return renderIterableBody(argument, ([key, value]) => [
      'li',
      {},
      reference(key),
      ': ',
      reference(value),
    ]);
  }
}

export class SetFormatter implements ChromeFormatter<Set<unknown>> {
  header(argument: Set<unknown>): Header | null {
    return isObservableSet(argument)
      ? renderIterableHeader(argument, 'Set')
      : null;
  }

  hasBody(argument: Set<unknown>): boolean {
    return argument.size > 0;
  }

  body(argument: Set<unknown>): Body {
    return renderIterableBody(argument, (value) => ['li', reference(value)]);
  }
}
