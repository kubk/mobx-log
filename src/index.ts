export { makeLoggable } from './make-loggable';
export { DefaultLogger } from './default-logger';
export type {
  ObservableEvent,
  ComputedEvent,
  ActionEvent,
  Logger,
} from './types';
export { DefaultLogWriter } from './log-writer';
export { configureMakeLoggable } from './config';
export { useLoggableLocalObservable } from './use-loggable-local-observable';
export { installMobxFormatters as applyFormatters } from './chrome-formatters';
