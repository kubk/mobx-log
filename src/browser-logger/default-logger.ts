import { ActionEvent, ComputedEvent, Logger, ObservableEvent } from '../types';
import { LogWriter } from './log-writer';

const colors = {
  red: '#FF6157',
  blue: '#28B7F6',
  orange: '#FFBC2D',
  green: '#2BC941',
};

export class DefaultLogger implements Logger {
  constructor(private logWriter: LogWriter) {}

  logObservable(event: ObservableEvent): void {
    if (event.type === 'update') {
      const info: unknown[] = [
        '%c[O]',
        `color:${colors.orange}`,
        event.name,
        event.oldValue,
        '->',
        event.newValue,
      ];

      this.logWriter.write(...info);
    }

    if (event.type === 'array') {
      const info: unknown[] = [
        '%c[O]',
        `color:${colors.orange}`,
        event.name,
        'has changed.',
      ];

      if (event.removed?.length) {
        info.push('Removed: ', event.removed);
      }
      if (event.added?.length) {
        info.push('Added: ', event.added);
      }

      this.logWriter.write(...info);
    }

    if (event.type === 'map') {
      const info: unknown[] = ['%c[O]', `color:${colors.orange}`, event.name];

      if (event.mapType === 'add') {
        info.push('- added', event.key, ':', event.value);
      }
      if (event.mapType === 'delete') {
        info.push('- deleted', event.key);
      }
      if (event.mapType === 'update') {
        info.push(
          '- updated',
          event.key,
          'value',
          event.oldValue,
          '->',
          event.value
        );
      }

      this.logWriter.write(...info);
    }

    if (event.type === 'set') {
      const info: unknown[] = ['%c[O]', `color:${colors.orange}`, event.name];

      if (event.setType === 'add') {
        info.push('- added', event.value);
      }
      if (event.setType === 'delete') {
        info.push('- removed', event.value);
      }

      this.logWriter.write(...info);
    }
  }

  logAction(event: ActionEvent): void {
    const formattedArguments =
      event.arguments.length > 0
        ? '%o, '.repeat(event.arguments.length).slice(0, -2)
        : '';

    const info: unknown[] = [
      `%c[A]%c ${event.name}(${formattedArguments})`,
      `color:${colors.red}`,
      'color:inherit',
      ...event.arguments,
    ];

    this.logWriter.write(...info);
  }

  logComputed(event: ComputedEvent): void {
    const info: unknown[] = [
      '%c[C]',
      `color:${colors.blue}`,
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }
}
