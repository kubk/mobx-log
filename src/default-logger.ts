import { ActionEvent, ComputedEvent, Logger, ObservableEvent } from './types';
import { LogWriter } from './log-writer';

const colors = {
  red: '#FF6157',
  blue: '#28B7F6',
  orange: '#FFBC2D',
  green: '#2BC941',
};

export class DefaultLogger implements Logger {
  constructor(private logWriter: LogWriter) {}

  logObservable(event: ObservableEvent) {
    const info = [
      '%c[O]',
      `color:${colors.orange}`,
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }

  logAction(event: ActionEvent) {
    const formattedArguments =
      event.arguments.length > 0
        ? '%o '.repeat(event.arguments.length).slice(0, -1)
        : '';

    const info: unknown[] = [
      `%c[A]%c ${event.name}(${formattedArguments})`,
      `color:${colors.red}`,
      'color:inherit',
      ...event.arguments,
    ];

    this.logWriter.write(...info);
  }

  logComputed(event: ComputedEvent) {
    const info = [
      '%c[C]',
      `color:${colors.blue}`,
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }
}
