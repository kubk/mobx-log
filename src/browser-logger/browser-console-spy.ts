import { comparer, spy } from 'mobx';
import type { PureSpyEvent } from 'mobx/dist/core/spy';
import { Logger } from '../types';
import { getStoreName, isStore } from '../store';
import { GlobalConfig } from '../global-config';

const parseDebugName = (name: string) => {
  return name.replace(/@\d+/, '');
};

export type StoreEventFilters = {
  computeds: boolean;
  observables: boolean;
  actions: boolean;
};

type FiltersPerStore = {
  [storeName in string]: StoreEventFilters;
};

export class BrowserConsoleSpy {
  private filtersPerStore: FiltersPerStore = {};

  constructor(
    private logger: Logger,
    private debug = false,
    private globalFilters: GlobalConfig['filters']
  ) {}

  listen() {
    spy((event: PureSpyEvent) => {
      const { logger } = this;

      if (this.debug) {
        console.log(event);
      }

      if (event.type === 'update') {
        if (event.observableKind === 'computed') {
          if (!this.globalFilters.computeds) {
            return;
          }
          const equals = comparer.default;
          if (!equals(event.oldValue, event.newValue)) {
            const computedFullName = parseDebugName(event.debugObjectName);
            const [storeName] = computedFullName.split('.');
            if (!this.filtersPerStore[storeName]?.computeds) {
              return;
            }
            logger.logComputed({
              name: computedFullName,
              oldValue: event.oldValue,
              newValue: event.newValue,
            });
          }
        }

        if (event.observableKind === 'object') {
          if (!this.globalFilters.observables) {
            return;
          }
          const observableFullName = parseDebugName(event.debugObjectName);
          const [storeName] = observableFullName.split('.');
          if (!this.filtersPerStore[storeName]?.observables) {
            return;
          }
          // Mobx spy adds '[..]' to object name in case change happened inside array
          const isArray = observableFullName.endsWith('[..]');
          const name = `${
            isArray ? observableFullName : storeName
          }.${event.name.toString()}`;

          logger.logObservable({
            type: 'update',
            name,
            newValue: event.newValue,
            oldValue: event.oldValue,
          });
        }
      }

      if (event.type === 'action') {
        if (!isStore(event.object) || !this.globalFilters.actions) {
          return;
        }
        const storeName = getStoreName(event.object);
        if (storeName === null || !this.filtersPerStore[storeName]?.actions) {
          return;
        }

        const action =
          'loggableName' in event.object
            ? event.name.split('.')[1]
            : event.name;

        logger.logAction({
          name: `${storeName}.${action}`,
          arguments: event.arguments,
        });
        return;
      }

      if ('observableKind' in event) {
        if (!this.globalFilters.observables) {
          return;
        }

        if (event.observableKind === 'array') {
          const observableFullName = parseDebugName(event.debugObjectName);
          const [storeName, key] = observableFullName.split('.');
          if (!this.filtersPerStore[storeName]?.observables) {
            return;
          }
          logger.logObservable({
            type: event.observableKind,
            name: `${storeName}.${key.toString()}`,
            // @ts-expect-error
            removed: event.removed,
            // @ts-expect-error
            added: event.added,
          });
        }

        if (event.observableKind === 'map') {
          const observableFullName = parseDebugName(event.debugObjectName);
          const [storeName, key] = observableFullName.split('.');
          if (!this.filtersPerStore[storeName]?.observables) {
            return;
          }
          logger.logObservable({
            type: 'map',
            key: event.name,
            value: 'newValue' in event ? event.newValue : undefined,
            oldValue: 'oldValue' in event ? event.oldValue : undefined,
            name: `${storeName}.${key.toString()}`,
            mapType: event.type,
          });
        }

        if (event.observableKind === 'set') {
          const observableFullName = parseDebugName(event.debugObjectName);
          const [storeName, key] = observableFullName.split('.');
          if (!this.filtersPerStore[storeName]?.observables) {
            return;
          }
          const value =
            event.type === 'delete'
              ? event.oldValue
              : event.type === 'add'
              ? event.newValue
              : undefined;

          logger.logObservable({
            type: 'set',
            value,
            name: `${storeName}.${key.toString()}`,
            setType: event.type,
          });
        }
      }
    });
  }

  addFilterByStore(name: string, storeEventFilters: StoreEventFilters) {
    this.filtersPerStore[name] = storeEventFilters;
  }
}
