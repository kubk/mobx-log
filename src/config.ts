import { DefaultLogger } from './browser-logger/default-logger';
import { Logger } from './types';
import { DefaultLogWriter } from './browser-logger/log-writer';

export enum LoggerType {
  ReduxDevtools = 'redux-devtools',
  BrowserConsole = 'browser-console',
}

export type Config = {
  logger: Logger;
  debug: boolean;
  type: LoggerType | null;
  storeConsoleAccess: boolean;
  filters: {
    computeds: boolean;
    actions: boolean;
    observables: boolean;
  };
};

export const config: Config = {
  logger: new DefaultLogger(new DefaultLogWriter()),
  type: null,
  debug: false,
  storeConsoleAccess: false,
  filters: {
    computeds: true,
    actions: true,
    observables: true,
  },
};

type DevToolsOptions = Partial<
  Pick<Config, 'debug' | 'storeConsoleAccess'>
>;

export const configureDevtools = (options?: DevToolsOptions) => {
  config.type = LoggerType.ReduxDevtools;
  if (options?.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options?.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
};

type LoggerOptions = Partial<
  Pick<Config, 'logger' | 'debug' | 'storeConsoleAccess' | 'filters'>
>;

export const configureLogger = (options?: LoggerOptions) => {
  config.type = LoggerType.BrowserConsole;
  if (options?.logger !== undefined) {
    config.logger = options.logger;
  }
  if (options?.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options?.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
  if (options?.filters !== undefined) {
    config.filters = options.filters;
  }
};
