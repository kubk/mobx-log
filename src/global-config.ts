import { DefaultLogger } from './browser-logger/default-logger';
import { Logger } from './types';
import { DefaultLogWriter } from './browser-logger/log-writer';

export enum LoggerType {
  ReduxDevtools = 'redux-devtools',
  BrowserConsole = 'browser-console',
}

export type GlobalConfig = {
  logger: Logger;
  debug: boolean;
  type: LoggerType | null;
  storeConsoleAccess: boolean;
  filters: {
    computeds: boolean;
    actions: boolean;
    observables: boolean;
  };
};

export const config: GlobalConfig = {
  logger: new DefaultLogger(new DefaultLogWriter()),
  type: null,
  debug: false,
  storeConsoleAccess: false,
  filters: {
    computeds: true,
    actions: true,
    observables: true,
  },
};

type DevToolsOptions = Partial<
  Pick<GlobalConfig, 'debug' | 'storeConsoleAccess'>
>;

export const configureDevtools = (options?: DevToolsOptions) => {
  config.type = LoggerType.ReduxDevtools;
  if (options?.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options?.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
};

type LoggerOptions = Partial<
  Pick<GlobalConfig, 'logger' | 'debug' | 'storeConsoleAccess' | 'filters'>
>;

export const configureLogger = (options?: LoggerOptions) => {
  config.type = LoggerType.BrowserConsole;
  if (options?.logger !== undefined) {
    config.logger = options.logger;
  }
  if (options?.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options?.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
  if (options?.filters !== undefined) {
    config.filters = options.filters;
  }
};
