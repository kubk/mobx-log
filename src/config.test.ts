import { autorun, makeAutoObservable } from 'mobx';
import { configureLogger, makeLoggable, DefaultLogger } from '../src';
import { CollectingLogWriter } from '../src/browser-logger/log-writer';

describe('config for production', () => {
  const previousEnv = process.env.NODE_ENV;
  process.env.NODE_ENV = 'production'

  const collectingWriter = new CollectingLogWriter();

  configureLogger({
    logger: new DefaultLogger(collectingWriter),
  });

  class StoreOnlyObservables {
    age?: number;
    name?: string;

    constructor() {
      makeAutoObservable(this);
      makeLoggable(this);
    }

    init() {
      this.age = 24;
      this.name = 'Test';
    }
  }

  it('does not log in production', () => {
    const storeOnlyObservables = new StoreOnlyObservables();
    storeOnlyObservables.init();

    expect(collectingWriter.history).toHaveLength(0);
    process.env.NODE_ENV = previousEnv;
  });
});

describe('config - filters', () => {
  const logWriter = new CollectingLogWriter();

  configureLogger({
    logger: new DefaultLogger(logWriter),
    storeConsoleAccess: true,
    filters: {
      observables: false,
      actions: false,
      computeds: true,
    },
  });

  class Counter {
    value = 0;

    constructor() {
      makeAutoObservable(this);
      makeLoggable(this);
    }

    increment() {
      this.value++;
    }

    get isEven() {
      return this.value % 2 === 0;
    }
  }

  it('respects filters', () => {
    const c = new Counter();

    let isEven;
    autorun(() => {
      isEven = c.isEven;
    });

    c.increment();
    c.increment();

    expect(logWriter.history).toMatchSnapshot();
  });
});

describe('config - console access', () => {
  configureLogger({
    logger: new DefaultLogger(new CollectingLogWriter()),
    storeConsoleAccess: true,
  });

  class Counter {
    value = 0;

    constructor() {
      makeAutoObservable(this);
      makeLoggable(this);
    }

    increment() {
      this.value++;
    }
  }

  class Todo {
    isDone = false;

    constructor() {
      makeAutoObservable(this);
      makeLoggable(this);
    }

    complete() {
      this.isDone = true;
    }
  }

  it('storeConsoleAccess', () => {
    new Counter();
    new Todo();

    expect(window.store).toBeTruthy();
    expect(window.store).toHaveProperty('counter');
    expect(window.store).toHaveProperty('todo');
    expect(typeof window.store!['todo'] === 'object').toBeTruthy();
    expect(Object.keys(window.store!)).toHaveLength(2);

    new Todo();

    expect(Object.keys(window.store!)).toHaveLength(2);
    expect(typeof window.store!['todo'] === 'object').toBeTruthy();
    expect(Object.keys(window.store!['todo'])).toHaveLength(1);
  });
})
