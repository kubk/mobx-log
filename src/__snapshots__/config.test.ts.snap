// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config - filters respects filters 1`] = `Array []`;
